### PR TITLE
[SOAR-4789] ICON Template Migration

### PR DESCRIPTION
**Pull request by jrose-r7**
## Description

* Convert the existing dataclass for workflow to use Pydantic - this is located in icon validators under workflow/model.py: https://github.com/rapid7/icon-integrations-validators/pull/138/files#diff-0d38e99c653c1c562d3da47237a7a4b5c81ca16b82b22b39766346cd55bc3f5dR8.
* Update icon-validate to take a flag to validate with template validators only: https://github.com/rapid7/icon-integrations-validators/pull/138/files#diff-299cf2bf9a4a9ea24829a813c5cfcc8baa77ae12002d41a167ea25eee499afaaR27
* Move validate_id and validate_workflow_version and workflowid_and_region to icon validators

Jira link: https://issues.corp.rapid7.com/browse/SOAR-4789

### CODE REVIEW

File: icon_validator/rules/__init__.py [hunk 1]
```diff
@@ -67,7 +67,15 @@
 from icon_validator.rules.workflow_validators.workflow_encoding_validator import *
 from icon_validator.rules.workflow_validators.workflow_parameters_keyword_validator import *
 
+# template imports
+from icon_validator.rules.template_validators.template_id_validator import *
+
 # The order of this list is the execution order of the validators.
+TEMPLATE_VALIDATORS = [
+    TemplateIDValidator()
```
Comment (lmilby-r7) [inline icon_validator/rules/__init__.py:75]: Think we are missing the validator for if the workflow ID exists. Thoughts?

File: icon_validator/workflow/model.py [hunk 2]
```diff
@@ -1,65 +1,167 @@
...
+from pydantic import BaseModel
 
 
 # Kom/Icon file
-@dataclass
-class TriggersInputActorProperties:
-    assets: dict = field(default_factory=lambda: {})
-    users: dict = field(default_factory=lambda: {})
+
+class TriggersOutputJsonSchemaPropertiesType(BaseModel):
```
Comment (pmara-r7) [inline icon_validator/workflow/model.py:8]: this is a nitpick but these class names are really long! Is the "JsonSchema" that is repeated across most of them needed?

File: icon_validator/rules/workflow_validators/workflow_id_validator.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+from typing import List
+from icon_validator.rules.validator import KomandPluginValidator
+from icon_validator.exceptions import ValidationException
+from icon_validator.workflow.unmarshal import read_workflow
+from icon_validator.workflow.model import Workflow
+from icon_validator.styling import YELLOW
+
+class WorkflowIDValidator(KomandPluginValidator):
+    def id_exists(self, workflow: Workflow):
+        if workflow.kom.workflowVersions[0].id:
```
Comment (pmara-r7) [inline icon_validator/rules/workflow_validators/workflow_id_validator.py:12]: This seems like a duplicate of the one above, did it not get changed in a copy/paste?

File: icon_validator/workflow/unmarshal.py [hunk 4]
```diff
@@ -16,5 +15,6 @@ def read_workflow(spec: dict, file_name: str) -> Workflow:
             wf = json.load(workflow_file)
         except Exception as e:
             raise Exception('Unable to read workflow', e)
-        return from_dict(data_class=Workflow, data=wf)
+        return parse_obj_as(Workflow, wf)
```
Comment (pmara-r7) [inline icon_validator/workflow/unmarshal.py:18]: Is it possible for this to throw exception?

**Review by pmara-r7 — COMMENTED** [review]
Looks clean overall, nice! Couple minor comments, beyond that, make sure this is well tested! There are a lot of pydantic models and want to test normal and corner cases here.

File: icon_validator/rules/template_validators/template_id_validator.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+from icon_validator.rules.validator import KomandPluginValidator
+from icon_validator.exceptions import ValidationException
+from icon_validator.workflow.unmarshal import read_workflow
+from icon_validator.workflow.model import Workflow
+from icon_validator.styling import YELLOW
+
+class TemplateIDValidator(KomandPluginValidator):
+
+    def id_exists(self, workflow: Workflow):
+        if workflow.kom.workflowVersions[0].id:
```
[the remark this answers is not in the document]
Comment (mrinehart-r7) [reply icon_validator/rules/template_validators/template_id_validator.py:13]: It should be pretty safe - if it is 0, then there are much bigger issues at play (within the product)

File: icon_validator/workflow/model.py (hunk 2 above)
Comment (jrose-r7) [reply icon_validator/workflow/model.py:8]: Good point! Wanted to try and differentiate between some overlapping attribute names but should make this more readable.

File: icon_validator/workflow/unmarshal.py [hunk 6]
```diff
@@ -1,8 +1,7 @@
 import os
 import json
 from icon_validator.workflow.model import Workflow
-from dacite import from_dict
-
+from pydantic import parse_obj_as
```
Comment (jrose-r7) [inline icon_validator/workflow/unmarshal.py:4]: since this was updated to use pydantic, should I change the dataclass reference below?

File: icon_validator/workflow/model.py [hunk 7]
```diff
@@ -1,65 +1,35 @@
 from typing import List, Optional, Type, Union
-from dataclasses import dataclass, field
+from dataclasses import field
```
[the remark this answers is not in the document]
Comment (jrose-r7) [reply icon_validator/workflow/model.py:2]: removed reference to field

File: icon_validator/rules/template_validators/template_id_validator.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+            return True
+        return False
+
+    def validate(self, spec):
+        """
+        Validates template ensuring ID
+        """
+        wf = Workflow
```
[the remark this answers is not in the document]
Comment (jrose-r7) [reply icon_validator/rules/template_validators/template_id_validator.py]: added instantiation

File: icon_validator/workflow/unmarshal.py (hunk 4 above)
Comment (jrose-r7) [reply icon_validator/workflow/unmarshal.py:29]: reference for parse_obj_as didn't specify an error as referenced here  https://github.com/samuelcolvin/pydantic/blob/master/pydantic/tools.py#L33 but I can put the parse_obj_as within the try block

File: icon_validator/rules/template_validators/template_workflow_version_validator.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+from icon_validator.rules.validator import KomandPluginValidator
+from icon_validator.exceptions import ValidationException
+from icon_validator.workflow.model import Workflow
+from icon_validator.workflow.unmarshal import read_workflow
+
+
+class TemplateWorkflowVersionValidator(KomandPluginValidator):
+
+    def version_exists(self, wf: Workflow) -> bool:
+        if wf.kom.workflowVersions[0].version:
```
Comment (pmara-r7) [inline icon_validator/rules/template_validators/template_workflow_version_validator.py:10]:
```suggestion
        if len(wf.kom.workflowVersions) > 0 and wf.kom.workflowVersions[0].version:
```

File: icon_validator/workflow/model.py [hunk 10]
```diff
@@ -1,65 +1,35 @@
...
-
-@dataclass
-class Trigger:
+class Trigger(BaseModel):
     id: str = ""
     name: str = ""
     description: str = ""
-    input: Optional[dict] = field(default_factory=lambda: {})
-    inputJsonSchema: TriggersInputJsonSchema = TriggersInputJsonSchema
-    outputJsonSchema: dict = field(default_factory=lambda: {})
```
[the remark this answers is not in the document]
Comment (pmara-r7) [reply icon_validator/workflow/model.py:41]: I'm not sure...Luke was def the most into pydantic

File: icon_validator/workflow/unmarshal.py [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    :return: Workflow dataclass
+    """
+
+    if not file_name:
+        directory = os.listdir(spec.directory)
+        for f in directory:
+            if f.endswith(".icon"):
+                file_name = f
+        if not file_name:
+            print("Not found")
```
Comment (pmara-r7) [inline icon_validator/workflow/unmarshal.py:22]: return here? If you don't have a filename, shouldn't go into the next "with" section

**Review by pmara-r7 — CHANGES_REQUESTED** [review]

**Review by pmara-r7 — APPROVED** [review]

**Review by mberezin-r7 — DISMISSED** [review]

File: Makefile [hunk 12]
```diff
@@ -11,5 +11,8 @@ disttest: package
 distprod: package
 	python3 -m twine upload dist/*
 
+localinstall:
+    pip3 install -e .
+
```
Comment (mrinehart-r7) [inline Makefile:16]:
```suggestion
localinstall:
      pip3 install -e .

```

File: README.md [hunk 13]
```diff
@@ -62,6 +62,7 @@ be on your way to contributing!
 
 ## Changelog
 
+* 2.41.0 - Convert dataclass to pydantic for Workflow to use pydantic | Update icon-validate to validate with template validators only | Move validate_id and validate_workflow_version to icon validators
```
Comment (mrinehart-r7) [inline README.md:65]:
```suggestion
* 2.41.0 - Convert dataclass to pydantic for Workflow to use pydantic | Update icon-validate to add template validation capabilities
```

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]
Looks good, however there is one last template validator that needs to be added: https://github.com/rapid7/icon-template/blob/master/icon_template/validators/validate_workflowid_and_region.py

File: setup.py [hunk 14]
```diff
@@ -7,7 +7,7 @@
 
 setup(
     name="insightconnect_integrations_validators",
-    version="2.40.0",
+    version="2.41.0",
```
Comment (mberezin-r7) [inline setup.py:10]:
```suggestion
    version="2.42.0",
```
Need to version bump after supported versions validator release.
